src/store/features/blocks/blocks.ts

```diff
@@ -0,0 +1,97 @@
+import tbConf, {ToolboxConfiguration} from "../../../toolboxconfig"
+import {produce} from 'immer'
+import {MutableRefObject} from "react";
+
+// Define actions
+const TOGGLE_CATEGORY = 'category/toggle'
+const TOGGLE_BLOCK_DESCRIPTION = 'block/toggleDescription'
+
+// Type actions
+type ToggleCategory = {
+  type: typeof TOGGLE_CATEGORY,
+  payload: {
+    category: number,
+    blocksDivRef: MutableRefObject<any>
+  }
+}
+
+type ToggleBlockDescription = {
+  type: typeof TOGGLE_BLOCK_DESCRIPTION,
+  payload: {
+    block: number,
+    blockRef: MutableRefObject<any>
+  }
+}
+
+// Action creators
+export const toggleCategoryAction = (categoryId: number, categoryRef: MutableRefObject<any>): ToggleCategory => ({
+  type: TOGGLE_CATEGORY,
+  payload: {
+    category: categoryId,
+    blocksDivRef: categoryRef
+  }
+})
+
+export const toggleBlockDescriptionAction = (blockId: number, blockRef: MutableRefObject<any>): ToggleBlockDescription => ({
+  type: TOGGLE_BLOCK_DESCRIPTION,
+  payload: {
+    block: blockId,
+    blockRef: blockRef
+  }
+})
+
+// Reducers
+type Actions = ToggleCategory | ToggleBlockDescription
+
+let calcHeight = true
+const blocksReducer = (state: ToolboxConfiguration = tbConf, action: Actions) => {
+  switch (action.type) {
+    case 'category/toggle':
+      // Calculate total height of all child block components for accordion animation
+      let maxHeight = 0;
+      (action.payload.blocksDivRef.current.childNodes as NodeListOf<HTMLDivElement>).forEach((block: HTMLDivElement) => {
+        maxHeight += block.clientHeight
+      })
+      // Update state
+      return produce(state, draftState => {
+        const foundCategory = draftState.categories.find(c => c.id === action.payload.category)
+        if (foundCategory) {
+          foundCategory.toggled = !foundCategory.toggled
+          foundCategory.toggled ? foundCategory.maxHeight = maxHeight : foundCategory.maxHeight = 0
```

Comment:
It seems to me that this `maxHeight` that you use is some design-related data can be derived and calculated from the values in the store. Thus, it should not be stored in the store. The store should be data-driven, it should only contain the data about the objects and their state. You can compute the maxHeight value in the render function of the appropriated component based on the store data (here the `toggled` state).

Reply:
Thank you for your explanation, I understand that the store should be data-driven and how this implementation was problematic. I have moved all related logic into useEffect hooks in the respective components.